text/0000-temp-components-lib.md

```diff
@@ -0,0 +1,618 @@
+
+function ToastProvider(props: PropsWithChildren<{}>) {
+  const [state, dispatch] = useReducer<ToastState>(reducer, initialToastState)
+
+  const value = useMemo(
+    () => ({
+      state,
+      dispatch
+    }),
+    [state, dispatch]
+  )
+
+  return (
+    <ToastContext.Provider value={value}>
+      {props.children}
+
+      {state.show &&
+        createPortal(
+          <Toast
+            onClose={() => closeToastAction(dispatch)}
+            sentiment={state.sentiment}
+            onUndo={state.onUndo}
+          >
+            <ToastHeading>{state.heading}</ToastHeading>
+            <p>{state.description}</p>
+          </Toast>
+        , document.body)}
+    </ToastContext.Provider>
+  )
+}
+
+interface ToastHookOptions extends ToastOptions {
+  heading?: string
+  description: string
+  onUndo?: () => {}
+}
+
+export function useToast() {
+  const toastRef = useRef(null)
+  const context = useContext(ToastContext)
+
+  if (!context) {
+    throw new Error(
+      'useToast requires a ToastProvider to be used on a parent component.'
+    )
+  }
+
+  const handleShow = useCallback((options: ToastState) => {
+    showToastAction(dispatch, options)
+
+    toastRef.current = setTimeout(() => {
+      closeToastAction(dispatch)
+    }, [9000])
+  }, [dispatch])
+
+  useEffect(() => {
+    return () => clearTimeout(toastRef.current)
+  }, [])
+
+  return useMemo(() => ({
+    close: closeToastAction(context.dispatch)
+    show: handleShow
+  }), [context, handleShow])
+}
+```
+
+</details>
+
+<br />
+
+In this example, it should allow the developer to have an high-performant
+experience like so (due to the use of useReducer):
+
+In a parent/route component:
+
+```typescript
+<ToastProvider>
+  <PageContent />
+</ToastProvider>
+```
+
+In any child of that route component:
+
+```typescript
+const toast = useToast()
+
+const handleUndo = useCallback(async () => {
+  // reverse user's action
+}, [])
+
+async function handleClick() {
+  try {
+    // perform some server update..
+    const update = await fetchThing()
+
+    toast.show({
+      sentiment: 'success',
+      title: 'Channels updated',
+      description: 'New channel bookmarked.'
+      onUndo: handleUndo
+    })
+  } catch (e) {
+    toast.show({
+      sentiment: 'danger',
+      title: 'Unable to update',
+      description: 'The server failed to update.
+    })
+  }
+}
+```
+
+With this design, the user only needs to:
+
+1. Add a Provider
+2. Use a hook that provides helper functions to show/hide a Toast notification without having to worry about managing a component or its state.
+
+Additionally, we should just re-export the `useTheme` hook since this library would
+make React-utils void for anyone using this library.
+
+### Helper Functions
+
+Along with all of the above, there should be some nice to have helpers that we ship
+that help make Pando React even better.
+
+I'm not totally for sure what these would be yet, but just noting.
```

Comment:
We created a `combineReducers` function which might be a good candidate 🙂 It's modeled after `combineReducers` from Redux and a couple other packages but we made it to be hooks-API-friendly.
- [combineReducers in our skill-assessments BC](https://gitlab.com/pluralsight/experience/learner/a-team/skill-assessments/-/blob/master/web/src/common/utils/combineReducers.js)
    - _for bonus points, [some tests](https://gitlab.com/pluralsight/experience/learner/a-team/skill-assessments/-/blob/master/web/src/common/utils/__tests__/combineReducers.test.js)_

Reply:
@saurookadook let me think on this one because that concept is technically an anti-pattern at this point and shouldn't be needed if components and React are used in the way it is intended (which is why Redux is dead). But, I want to make sure you hear that I don't mind adding it if it makes sense, I just am not sure it does for this library.

I also can't view the link because I don't have GL in my Okta. 😢

Reply:
Oh bummer! I'll Slack you the code for the utility just so you can have a look; the big gain we've found from the utility is being able to more easily organize our application state. (_You might actually be able to surmise that from the mocks in the test_ 🙂 )

Reply:
> (which is why Redux is dead).

Ooh boy that's a bold statement 😅 Having worked a little bit with the current maintainer of redux and contributed to `@reduxjs/toolkit`, I feel obligated to say that modern redux is actually quite nice IMHO (e.g. using `immer` for reducers, much less boilerplate than before).

Idk what @saurookadook is linking to bc I can't access GL either haha, but if what they're referring to is a way to combine the data in multiple `useReducer` calls, that really has nothing to with redux and could potentially make sense as a utility in `react-utils` if it's a common use case.

Reply:
@josh-degraw exactly and that's what we talked about in a private DM. Totally forgot to put it here but am glad you are on top of things! 👀